File: mentibot/mentibot.py
import requests
import json

def get_vote_key(code: str) -> str:

	headers = {
		'authority': 'www.menti.com',
		'sec-ch-ua': '"Chromium";v="93", " Not;A Brand";v="99"',
		'accept': 'application/json',
		'content-type': 'application/json; charset=UTF-8',
		'sec-ch-ua-mobile': '?0',
		'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36',
		'sec-ch-ua-platform': '"Windows"',
		'sec-fetch-site': 'same-origin',
		'sec-fetch-mode': 'cors',
		'sec-fetch-dest': 'empty',
		'referer': 'https://www.menti.com/',
		'accept-language': 'de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7',
	}

	r = requests.get(f"https://www.menti.com/core/vote-ids/{code}/series", headers=headers).json()
	
	return r

def get_public_key(vote_key: str) -> str:

	headers = {
		'authority': 'www.menti.com',
		'sec-ch-ua': '";Not A Brand";v="99", "Chromium";v="94"',
		'accept': 'application/json',
		'content-type': 'application/json; charset=UTF-8',
		'sec-ch-ua-mobile': '?0',
		'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36',
		'sec-ch-ua-platform': '"Windows"',
		'sec-fetch-site': 'same-origin',
		'sec-fetch-mode': 'cors',
		'sec-fetch-dest': 'empty',
		'referer': f'https://www.menti.com/{vote_key}',
		'accept-language': 'de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7',
	}

	r = requests.get(f"https://www.menti.com/core/vote-keys/{vote_key}/series", headers=headers)
	
	return r.json()

def get_identifier() -> str:
	headers = {
	"user-agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 10_0 like Mac OS X) AppleWebKit/602.1.50 (KHTML, like Gecko) Version/10.0 YaBrowser/17.4.3.195.10 Mobile/14A346 Safari/E7FBAF"
	}
	
	r = requests.post("https://www.menti.com/core/identifier", headers = headers)
	
	return r.json()

def make_vote(vote_key: str, public_key: str, identifier: str, words: str):

	headers = {
    'authority': 'www.menti.com',
    'sec-ch-ua': '";Not A Brand";v="99", "Chromium";v="94"',
    'accept': 'application/json',
    'content-type': 'application/json; charset=UTF-8',
    'sec-ch-ua-mobile': '?0',
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36',
    'sec-ch-ua-platform': '"Windows"',
    'origin': 'https://www.menti.com',
    'sec-fetch-site': 'same-origin',
    'sec-fetch-mode': 'cors',
    'sec-fetch-dest': 'empty',
    'referer': f'https://www.menti.com/{vote_key}',
	'x-identifier':identifier,
    'accept-language': 'de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7',
	}
	#vote = " ".join([option1, option2, option3])
	data = '{"question_type":"wordcloud","vote":"%s"}' % words
	
	response = requests.post(f'https://www.menti.com/core/votes/{public_key}', headers=headers, data=data)
	if response.status_code == 200:
		print(f"[+] Successfully added words to the cloud ({words})")
	else:
		print("[-] Failed adding words to the cloud")
		
def add_words(code: str, vote_key: str, public_key: str, words: str) -> None:

	
	identifier = get_identifier()['identifier']
	#print(f"\n[+] Identifier: {identifier}")
	
	make_vote(vote_key, public_key, identifier, words)
	
	
if __name__ == '__main__':
	
	
	code = input("[?] Enter the code: ").replace(' ', '')
	file = input("[?] Filename of the file with the words you want to add: ")
	
	vote_key = get_vote_key(code)['vote_key']
	print(f"\n[*] Vote Key: {vote_key}")
	public_key = get_public_key(vote_key)['questions'][0]['public_key']
	print(f"[*] Public Key: {public_key}\n")
	
	with open(file, 'r') as f:
		total_words = f.read().splitlines()
	for i in range(0, len(total_words)-2, 3):
		words = f"{total_words[i]} {total_words[i+1]} {total_words[i+2]}"
		add_words(code, vote_key, public_key, words)
	
	
	

File: mentibot/mentibot.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strconv"
	"strings"
)

func main() {

	var code string
	flag.StringVar(&code, "c", "", "the code youre supposed to enter on menti.com")

	var wordfile string
	flag.StringVar(&wordfile, "w", "words.txt", "file that contains the words to be added")

	var repeat int
	flag.IntVar(&repeat, "r", 1, "how often to loop through the file")

	flag.Parse()

	var words_file []string

	f, _ := os.Open(wordfile)
	defer f.Close()
	var scanner = bufio.NewScanner(f)

	for scanner.Scan() {
		words_file = append(words_file, scanner.Text())
	}

	var words_vote []string

	min := func(a, b int) int {
		if a <= b {
			return a
		}
		return b
	}

	var vote_key = getVoteKey(code)
	var public_key, entries = getPublicKeyAndEntries(vote_key)

	for i := 0; i < len(words_file); i += entries {
		batch := words_file[i:min(i+entries, len(words_file))]
		words_vote = append(words_vote, strings.Join(batch, " "))
	}

	for i := 0; i < repeat; i++ {
		for _, words := range words_vote {
			var identifier = getIdentifier()
			makeVote(vote_key, public_key, identifier, words)
		}
	}
}

func getVoteKey(code string) string {

	client := &http.Client{}
	request, _ := http.NewRequest("GET", fmt.Sprintf("https://www.menti.com/core/vote-ids/%s/series", code), nil)

	request.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36")
	request.Header.Set("Content-Type", "application/json; charset=UTF-8")
	request.Header.Set("Referer", "https://www.menti.com/")
	request.Header.Set("Authority", "www.menti.com")

	response, err := client.Do(request)
	if err != nil {
		fmt.Println("Failed making request")
		os.Exit(0)
	}

	data, _ := ioutil.ReadAll(response.Body)
	defer response.Body.Close()

	var voteKeyJson map[string]string

	json.Unmarshal([]byte(data), &voteKeyJson)

	return voteKeyJson["vote_key"]

}

func getPublicKeyAndEntries(vote_key string) (string, int) {

	client := &http.Client{}
	request, _ := http.NewRequest("GET", fmt.Sprintf("https://www.menti.com/core/vote-keys/%s/series", vote_key), nil)

	request.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36")
	request.Header.Set("Content-Type", "application/json; charset=UTF-8")
	request.Header.Set("Referer", fmt.Sprintf("https://www.menti.com/%s", vote_key))
	request.Header.Set("Authority", "www.menti.com")

	response, err := client.Do(request)
	if err != nil {
		fmt.Println("Failed making request")
		os.Exit(0)
	}

	data, _ := ioutil.ReadAll(response.Body)
	defer response.Body.Close()
	//fmt.Println(string(data))
	var publicKeyJson map[string][]map[string]string

	json.Unmarshal([]byte(data), &publicKeyJson)
	entries, _ := strconv.Atoi(publicKeyJson["questions"][0]["max_nb_words"])
	publicKey := publicKeyJson["questions"][0]["public_key"]
	return publicKey, entries
}

func getIdentifier() string {
	client := &http.Client{}
	request, _ := http.NewRequest("POST", "https://www.menti.com/core/identifier", nil)

	request.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36")

	response, err := client.Do(request)
	if err != nil {
		fmt.Println("Failed making request")
		os.Exit(0)
	}

	data, _ := ioutil.ReadAll(response.Body)
	defer response.Body.Close()

	var identifierJson map[string]string

	json.Unmarshal([]byte(data), &identifierJson)

	return identifierJson["identifier"]
}

func makeVote(vote_key string, public_key string, identifier string, words string) {

	var jsonStr = []byte(fmt.Sprintf(`{"question_type":"wordcloud","vote":"%s"}`, words))

	client := &http.Client{}
	request, _ := http.NewRequest("POST", fmt.Sprintf("https://www.menti.com/core/votes/%s", public_key), bytes.NewBuffer(jsonStr))

	request.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36")
	request.Header.Set("Content-Type", "application/json; charset=UTF-8")
	request.Header.Set("Referer", fmt.Sprintf("https://www.menti.com/%s", vote_key))
	request.Header.Set("Authority", "www.menti.com")
	request.Header.Set("X-Identifier", identifier)

	_, err := client.Do(request)
	if err != nil {
		fmt.Println("[-] Failed adding words")
	} else {
		fmt.Printf("[+] Added words to the wordcloud (%s)\n", words)
	}
}
